### PR TITLE
Add unit tests for io.openmessaging.storage.dledger.utils.DLedgerUtils

### PR DESCRIPTION
**Pull request by JohnLBergqvist**
I've analysed your codebase and noticed that `io.openmessaging.storage.dledger.utils.DLedgerUtils` is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

### CODE REVIEW

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import org.powermock.modules.junit4.PowerMockRunner;
+
+@RunWith(PowerMockRunner.class)
+public class DLedgerUtilsTest {
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+  
+  @PrepareForTest({DLedgerUtils.class, System.class})
+  @Test
+  public void testComputNextTimeMillis() throws Exception {
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: Exception can be removed

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    PowerMockito.when(System.currentTimeMillis()).thenReturn(1515585600000L);
+
+    assertEquals(1515628800000L, DLedgerUtils.computNextMorningTimeMillis());
+    assertEquals(1515585660000L, DLedgerUtils.computNextMinutesTimeMillis());
+    assertEquals(1515589200000L, DLedgerUtils.computNextHourTimeMillis());
+    assertEquals(1515591000000L, DLedgerUtils.computNextHalfHourTimeMillis());
+  }
+
+  @Test
+  public void testCrc32() throws Exception {
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: Exception can be removed

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  @Test
+  public void testString2bytes() {
+    assertNull(DLedgerUtils.string2bytes(""));
+    final byte[] expected = new byte[] {-1};
+    assertArrayEquals(expected, DLedgerUtils.string2bytes(")@"));
+  }
+
+  @Test
+  public void testUncompress() throws Exception, IOException {
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: IOException can be removed as Exception is thrown

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    final byte[] src = {120, 1, 1, 3, 0, -4, -1, 0, 0, 0, 0, 3, 0, 1};
+    assertArrayEquals(new byte[] {0, 0, 0}, DLedgerUtils.uncompress(src));
+  }
+
+  @Test
+  public void testCompress() throws Exception, IOException {
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: IOException can be removed as Exception is thrown

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+ */
+
+package io.openmessaging.storage.dledger.util;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: Import not used `import static org.powermock.api.mockito.PowerMockito.mockStatic;`

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    assertEquals(1515628800000L, DLedgerUtils.computNextMorningTimeMillis());
+    assertEquals(1515585660000L, DLedgerUtils.computNextMinutesTimeMillis());
+    assertEquals(1515589200000L, DLedgerUtils.computNextHourTimeMillis());
+    assertEquals(1515591000000L, DLedgerUtils.computNextHalfHourTimeMillis());
+  }
+
+  @Test
+  public void testCrc32() throws Exception {
+    final byte[] nullArray = null;
+    assertEquals(0, DLedgerUtils.crc32(nullArray));
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: `assertEquals(0, DLedgerUtils.crc32(nullArray));` can be written as `assertEquals(0, DLedgerUtils.crc32(null));` without the nullArray

This is personal preference for me but you can save a few lines :)
There are other lines where I would do this as well :)

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    final byte[] array = {};
+    assertEquals(0, DLedgerUtils.crc32(array));
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: Same for this with the byte[] parameter

**Review by EricHetti — CHANGES_REQUESTED** [review]

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  @Test
+  public void testCrc32() {
+    assertEquals(0, DLedgerUtils.crc32(null));
+    assertEquals(0, DLedgerUtils.crc32(new byte[]{}));
+  }
+
+  @Test
+  public void testBytes2string() {
+    final byte[] src = {-104};
+    assertEquals("98", DLedgerUtils.bytes2string(src));
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: You missed one here

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  @Test
+  public void testCompress() throws IOException {
+    assertArrayEquals(
+        new byte[] {120, 1, 1, 3, 0, -4, -1, 0, 0, 0, 0, 3, 0, 1},
+        DLedgerUtils.compress(new byte[]{0, 0, 0}, 0)
+      );
+  }
+
+  @Test
+  public void testfrontStringAtLeast() {
```
Comment (Braavos6) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: uppercase F `testfrontStringAtLeast()`

**Review by EricHetti — APPROVED** [review]

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    assertEquals(1515628800000L, DLedgerUtils.computNextMorningTimeMillis());
+    assertEquals(1515585660000L, DLedgerUtils.computNextMinutesTimeMillis());
+    assertEquals(1515589200000L, DLedgerUtils.computNextHourTimeMillis());
+    assertEquals(1515591000000L, DLedgerUtils.computNextHalfHourTimeMillis());
+  }
+
+  @Test
+  public void testCrc32() {
+    assertEquals(0, DLedgerUtils.crc32(null));
+    assertEquals(0, DLedgerUtils.crc32(new byte[]{}));
```
Comment (freddytuxworth) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java:56]: add test with non-empty array

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  @Test
+  public void testBytes2string() {
+    assertEquals("98", DLedgerUtils.bytes2string(new byte[]{-104}));
```
Comment (freddytuxworth) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java]: This is a bit confusing IMO, maybe put `(byte) 0x98` instead of `-104`.
Also could do with a test with array length >1

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  @Test
+  public void testString2bytes() {
+    assertNull(DLedgerUtils.string2bytes(""));
+    assertArrayEquals(new byte[] {-1}, DLedgerUtils.string2bytes(")@"));
```
Comment (freddytuxworth) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java:70]: This is unexpected/undefined behaviour - could add a test where the string is an even number of hex chars

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  public void testUncompress() throws IOException {
+    final byte[] src = {120, 1, 1, 3, 0, -4, -1, 0, 0, 0, 0, 3, 0, 1};
+    assertArrayEquals(new byte[] {0, 0, 0}, DLedgerUtils.uncompress(src));
+  }
+
+  @Test
+  public void testCompress() throws IOException {
+    assertArrayEquals(
+        new byte[] {120, 1, 1, 3, 0, -4, -1, 0, 0, 0, 0, 3, 0, 1},
+        DLedgerUtils.compress(new byte[]{0, 0, 0}, 0)
```
Comment (freddytuxworth) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java:84]: excellent compression ratio :laughing: test is good though

File: src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+  @Test
+  public void testIsInternalIP() {
+    assertFalse(DLedgerUtils.isInternalIP(new byte[]{0, 0, 0, 0}));
+    assertTrue(DLedgerUtils.isInternalIP(new byte[]{10, 80, 0, 0}));
+    assertTrue(DLedgerUtils.isInternalIP(new byte[]{-84, 24, 0, 0}));
+    assertFalse(DLedgerUtils.isInternalIP(new byte[]{-84, 80, 0, 0}));
+    assertFalse(DLedgerUtils.isInternalIP(new byte[]{-84, 0, 0, 0}));
+    assertTrue(DLedgerUtils.isInternalIP(new byte[]{-64, -88, 0, 0}));
+    assertFalse(DLedgerUtils.isInternalIP(new byte[]{-64, 80, 0, 0}));
```
Comment (freddytuxworth) [inline src/test/java/io/openmessaging/storage/dledger/util/DLedgerUtilsTest.java:116]: Maybe split out the `False`s and the `True`s? no biggie though

**Review by freddytuxworth — CHANGES_REQUESTED** [review]

**Review by freddytuxworth — APPROVED** [review]
